pkg/didcomm/protocol/outofband/service.go

```diff
@@ -0,0 +1,341 @@
+	connID, err := s.didSvc.HandleInbound(c.msg, c.myDID, c.theirDID)
+	if err != nil {
+		return fmt.Errorf("didexchange service failed to handle inbound request : %w", err)
+	}
+
+	req := &Request{}
+
+	err = c.msg.Decode(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request message : %w", err)
+	}
+
+	// TODO if we want to implement retries then we should be saving state before invoking
+	//  the didexchange service
+	err = s.save(&myState{
+		ID:           c.msg.ID(),
+		ConnectionID: connID,
+		Request:      req,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to save my state : %w", err)
+	}
+
+	return nil
+}
+
+func (s *Service) handleDidEvent(e service.StateMsg) error {
+	// TODO remove 'empty parent threadID check'?
+	if e.Type != service.PostState || e.Msg.Type() != didexchange.AckMsgType || e.Msg.ParentThreadID() == "" {
+		// we are only interested in a successfully completed didexchange.
+		// the out-of-band protocol thread should be the did-exchange's parent thread.
+		return errIgnoredDidEvent
+	}
+
+	state, err := s.fetchMyState(e.Msg.ParentThreadID())
+	if err != nil {
+		return fmt.Errorf("failed to load state data with id=%s : %w", e.Msg.ParentThreadID(), err)
+	}
+
+	req, found := s.getNextRequestFunc(state)
+	if !found {
+		return errIgnoredDidEvent
+	}
+
+	bytes, err := s.extractDIDCommMsgBytesFunc(req)
+	if err != nil {
+		return fmt.Errorf("failed to extract didcomm message from attachment : %w", err)
+	}
+
+	record, err := s.fetchConnectionRecord(state.ConnectionID)
+	if err != nil {
+		return fmt.Errorf("failed to fetch connection record with id=%s : %w", state.ConnectionID, err)
+	}
+
+	err = s.dispatch(bytes, record.MyDID, record.TheirDID)
+	if err != nil {
+		return fmt.Errorf("failed to dispatch message : %w", err)
+	}
+
+	// TODO do we need the capability to register for events from whatever protocol service is handling that msg?
+
+	// TODO we're only processing a single message for now
+	state.Done = true
+
+	err = s.save(state)
+	if err != nil {
+		return fmt.Errorf("failed to update state : %w", err)
+	}
+
+	return nil
+}
+
+func (s *Service) save(state *myState) error {
+	bytes, err := json.Marshal(state)
+	if err != nil {
+		return fmt.Errorf("failed to save state=%+v : %w", state, err)
+	}
+
+	err = s.store.Put(state.ID, bytes)
+	if err != nil {
+		return fmt.Errorf("failed to save state : %w", err)
+	}
+
+	return nil
+}
+
+func (s *Service) fetchMyState(id string) (*myState, error) {
+	bytes, err := s.store.Get(id)
+	if err != nil && !errors.Is(err, storage.ErrDataNotFound) {
+		return nil, fmt.Errorf("failed to fetch state data with id=%s : %w", id, err)
+	}
+
+	if errors.Is(err, storage.ErrDataNotFound) {
+		// this didexchange thread did not spawn from an out-of-band thread that we know of
+		return nil, fmt.Errorf("cannot find a state record for pthid=%s : %w", id, err)
+	}
+
```

Comment:
I don't see any reason to have different error messages. Why do we need this?

Reply:
I cannot give you a definitive answer. 

I CAN explain however that those two error-checks we initially inside `handleDidEvent()`, where their intention would be perhaps more clear. (I moved them in here to please the cyclo linter).

If you imagine they're still in `handleDidEvent()` then see that:

1. the store may throw a generic error (normal)
2. no state may be found for the given pthid

I believe all did-exchange threads should spawn from this new out-of-band protocol. Therefore the incoming didexchange msg SHOULD have a pthid known to this service. Hence why I differentiated the errors between (1) and (2).

Make sense?

Reply:
I would not do that. For me, it looks complicated.

Reply:
Ok - removed (one less test case!)